### PR TITLE
Fuse batchnorm allreduces in FP/BP

### PR DESCRIPTION
**Pull request by ndryden**
This replaces the separate mean and variance (and associated gradients) buffers with fused buffers. This enables the allreduces during forward/backprop when doing node-local/global statistics to be fused into a single allreduce. Since the size of these buffers is the number of channels in the layer being normalized, these allreduces are likely latency-bound, so doing fewer of them improves performance.

Tested that it could train Wide ResNet-50-4 with node-local stats to convergence without issue. Also tested gradient checking.

Benchmarked with global stats at small scale and observed a minor but consistent performance improvement.

### CODE REVIEW

**Comment by ndryden** [comment]
I actually thought about that, but it doesn't work without more significant changes. The basic problem is that the views would be a `CPUMat` or a `GPUMat` (since H doesn't support `AbsMat` views), but our non-training code path expects `AbsMat&`s.

**Comment by timmoon10** [comment]
Ah... unfortunately that's true. Ok, this is the right approach then.

**Comment by bvanessen** [comment]
Sounds great.


Brian C. Van Essen

vanessen1@llnl.gov<mailto:vanessen1@llnl.gov>

(w) 925-422-9300

(c) 925-290-5470

---

Sent from my iPhone

On Jul 17, 2019, at 4:04 PM, Tim Moon <notifications@github.com<mailto:notifications@github.com>> wrote:


Ah... unfortunately that's true. Ok, this is the right approach then.
